### PR TITLE
【PaddlePaddle Hackathon 2】16 新增 API RRelu

### PR DESCRIPTION
**Pull request by thunder95**
### PR types
New features

### PR changes
APIs

### Describe
完成第二期第16项目开发任务: https://github.com/PaddlePaddle/Paddle/issues/40317
RRELU激活函数是从[Empirical Evaluation of Rectified Activations in Convolutional Network](https://arxiv.org/abs/1505.00853)中提出的，它是在Leaky ReLU基础上，对每一个位置的负值输入线性项做了随机采样 ，来加强一定范围内的泛化能力。
RFC设计文档: https://github.com/PaddlePaddle/community/pull/71
中文文档: https://github.com/PaddlePaddle/docs/pull/4725

### CODE REVIEW

**Comment by dingjiaweiww** [comment]
请先通过CI噢~

**Comment by SmirnovKol** [comment]
I think the implementation should  reference  the  implementation of  dropout.  Your op  attributes does not include "seed" ,  "fix_seed" , "is_test".

**Comment by SmirnovKol** [comment]
I think the implementation should  reference  the  implementation of  dropout.  Your op  attributes does not include "seed" ,  "fix_seed" , "is_test".

File: paddle/fluid/operators/rrelu_op.cc [hunk 1]
```diff
@@ -0,0 +1,153 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserved.
```
Comment (zhiboniu) [inline paddle/fluid/operators/rrelu_op.cc]: 2016->2022，其他文件自查一下吧

File: python/paddle/nn/layer/activation.py [hunk 2]
```diff
@@ -436,6 +436,74 @@ def extra_repr(self):
...
+
+            data = np.array([[[[-2.0,  3.0, -4.0,  5.0],
+                            [ 3.0, -4.0,  5.0, -6.0],
+                            [-7.0, -8.0,  8.0,  9.0]],
+                            [[ 1.0, -2.0, -3.0,  4.0],
+                            [-5.0,  6.0,  7.0, -8.0],
+                            [ 6.0,  7.0,  8.0,  9.0]]]], 'float64')
+            x = paddle.to_tensor(data)
+            m = paddle.nn.RReLU(0.1, 0.3)
+            out = m(x)
```
Comment (zhiboniu) [inline python/paddle/nn/layer/activation.py]: 示例代码需要正式一点，不能用m这种随意的名字，最好名字能看出来表示的功能

File: paddle/fluid/operators/rrelu_op.cc [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+)DOC");
+  }
+};
+
+class RReluGradOp : public framework::OperatorWithKernel {
+ public:
+  using framework::OperatorWithKernel::OperatorWithKernel;
+
+  void InferShape(framework::InferShapeContext* ctx) const override {
```
Comment (zhiboniu) [inline paddle/fluid/operators/rrelu_op.cc]: 正向的infershape已经放到phi里了，反向的infershape也一起放过去吧

File: paddle/phi/kernels/cpu/rrelu_kernel.cc [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+  const T* x_ptr = x.data<T>();
+  T* o_ptr = dev_ctx.template Alloc<T>(out);
+  T* n_ptr = dev_ctx.template Alloc<T>(noise);
+  T zero = static_cast<T>(0);
+  int numel = x.numel();
+  int i = 0;
+
+  if (is_test) {
+    for (i = 0; i < numel; i++) {
+      T mid_val = static_cast<T>((lower + upper) / 2.0);
```
Comment (zhiboniu) [inline paddle/phi/kernels/cpu/rrelu_kernel.cc]: 这个mid_val计算可以放到for循环外边吧

File: paddle/phi/kernels/gpu/rrelu_kernel.cu [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+}  // namespace phi
+
+PD_REGISTER_KERNEL(rrelu,
+                   GPU,
+                   ALL_LAYOUT,
+                   phi::RReluKernel,
+                   float,
+                   phi::dtype::float16,
```
Comment (zhiboniu) [inline paddle/phi/kernels/gpu/rrelu_kernel.cu:111]: 这个float16 cpu kernel也注册上吧

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+        self.upper = 0.3
+
+    def init_dtype(self):
+        self.dtype = "float64"
+
+    def init_input_shape(self):
+        self.x_shape = [2, 3, 4, 5]
+
+    def init_attr(self):
+        self.attrs = {'lower': self.lower, "upper": self.upper, "is_test": True}
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_rrelu_op.py]: is_test: False的情况也需要测试

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py [hunk 7]
```diff
@@ -0,0 +1,305 @@
...
+    def test_static_graph_functional(self):
+        '''test_static_graph_functional'''
+
+        for place in self.places:
+            paddle.enable_static()
+            x_1 = paddle.fluid.data(
+                name="x", shape=self.x_np.shape, dtype="float64")
+            x_2 = paddle.fluid.data(
+                name="x2", shape=self.x_np.shape, dtype="float64")
+            out_1 = F.rrelu(x_1, self.lower_0, self.upper_0, training=False)
```
Comment (zhiboniu) [inline python/paddle/fluid/tests/unittests/test_rrelu_op.py:96]: training=True的情况也需要测试

Comment (thunder95) [reply python/paddle/fluid/tests/unittests/test_rrelu_op.py:96]: @zhiboniu  training的时候noise和输出都是随机的，然而在optest中的输出都是固定的，请问老师有什么建议？可以将lower和upper设置成一样的值么？

Comment (zhiboniu) [reply python/paddle/fluid/tests/unittests/test_rrelu_op.py:96]: 可以试一下固定seed结果能不能固定。如果还是不行我看你现在检测范围是否在[lower*x, upper*x]也可以。

File: paddle/fluid/operators/rrelu_op.cc (hunk 1 above)
Comment (thunder95) [reply paddle/fluid/operators/rrelu_op.cc]: 完成

File: python/paddle/nn/layer/activation.py (hunk 2 above)
Comment (thunder95) [reply python/paddle/nn/layer/activation.py]: 完成

File: paddle/fluid/operators/rrelu_op.cc (hunk 3 above)
Comment (thunder95) [reply paddle/fluid/operators/rrelu_op.cc]: 完成

File: paddle/phi/kernels/cpu/rrelu_kernel.cc (hunk 4 above)
Comment (thunder95) [reply paddle/phi/kernels/cpu/rrelu_kernel.cc]: 完成

File: paddle/phi/kernels/gpu/rrelu_kernel.cu (hunk 5 above)
Comment (thunder95) [reply paddle/phi/kernels/gpu/rrelu_kernel.cu:111]: 完成

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py (hunk 6 above)
Comment (thunder95) [reply python/paddle/fluid/tests/unittests/test_rrelu_op.py]: 完成，因opt_test的输出校验在1e-7范围内而noise是随机的，所以将lower和upper差距调到尽可能小来测试。

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py (hunk 7 above)
Comment (thunder95) [reply python/paddle/fluid/tests/unittests/test_rrelu_op.py:96]: 完成

**Review by zhiboniu — DISMISSED** [review]

**Comment by zhiboniu** [comment]
请注意补充中文文档啊

**Comment by zhiboniu** [comment]
随机数改成调封装好的函数吧（具体可以参考现有调用的代码）：
    funcs::uniform_distribution<T> dist;
    funcs::uniform_real_transform<T> trans(min, max);
    funcs::distribution_and_transform<T>(dev_ctx, out, dist, trans);

**Review by zhwesky2010 — COMMENTED** [review]
1.`fix_seed`这个参数应该不需要，`seed` 也不需要，目前的2.0后的写法是直接从generator中获取seed，不需要单独去给OP设置一个seed。
```
auto gen_cuda = ctx.GetGenerator();
uint64_t seed = seed_offset.first;
uint64_t offset = seed_offset.second;
```

2.GPU kernel 参考下 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/exponential_op.cu#L29-L31 中的简洁写法：
```
using MT = typename kps::details::MPTypeTrait<T>::Type;
funcs::uniform_distribution<MT> dist;
funcs::uniform_real_transform<MT> trans(min, max);
funcs::distribution_and_transform<T>(dev_ctx, out, dist, trans);
```
调用https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/funcs/distribution_helper.h 已经封装好的函数即可。不要用thrust库，这个性能不太好。可以参考下https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/gpu/poisson_kernel.cu 比较新的写法

**Comment by thunder95** [comment]
> 请注意补充中文文档啊

@[zhiboniu](https://github.com/zhiboniu) 已补充， 但未通过CI， 不知为什么会参数名称校验失败

**Comment by thunder95** [comment]
> 随机数改成调封装好的函数吧（具体可以参考现有调用的代码）： funcs::uniform_distribution dist; funcs::uniform_real_transform trans(min, max); funcs::distribution_and_transform(dev_ctx, out, dist, trans);

已修改 @zhiboniu

**Comment by thunder95** [comment]
> 1.`fix_seed`这个参数应该不需要，`seed` 也不需要，目前的2.0后的写法是直接从generator中获取seed，不需要单独去给OP设置一个seed。
> 
> ```
> auto gen_cuda = ctx.GetGenerator();
> uint64_t seed = seed_offset.first;
> uint64_t offset = seed_offset.second;
> ```
> 
> 2.GPU kernel 参考下 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/exponential_op.cu#L29-L31 中的简洁写法：
> 
> ```
> using MT = typename kps::details::MPTypeTrait<T>::Type;
> funcs::uniform_distribution<MT> dist;
> funcs::uniform_real_transform<MT> trans(min, max);
> funcs::distribution_and_transform<T>(dev_ctx, out, dist, trans);
> ```
> 
> 调用https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/funcs/distribution_helper.h 已经封装好的函数即可。不要用thrust库，这个性能不太好。可以参考下https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/gpu/poisson_kernel.cu 比较新的写法

@zhouwei25 谢谢老师指导，已完成修改意见，麻烦看看这样改是否可以。

File: python/paddle/nn/layer/activation.py [hunk 8]
```diff
@@ -436,6 +436,75 @@ def extra_repr(self):
             name_str)
 
 
+class RReLU(Layer):
+    """
```
Comment (Ligoml) [inline python/paddle/nn/layer/activation.py]: image

格式混乱，请参考[API文档规范](https://github.com/PaddlePaddle/docs/wiki/%E9%A3%9E%E6%A1%A8API%E6%96%87%E6%A1%A3%E4%B9%A6%E5%86%99%E8%A7%84%E8%8C%83)修改

File: python/paddle/nn/functional/activation.py [hunk 9]
```diff
@@ -548,6 +548,102 @@ def prelu(x, weight, data_format="NCHW", name=None):
     return out
 
 
+def rrelu(x, lower=1. / 8., upper=1. / 3., training=True, name=None):
+    """
```
Comment (Ligoml) [inline python/paddle/nn/functional/activation.py]: image

一样的，格式混乱

File: python/paddle/nn/layer/activation.py [hunk 10]
```diff
@@ -436,6 +436,75 @@ def extra_repr(self):
...
+    Shape:
+        - input: Tensor with any shape. Default dtype is float32.
+        - output: Tensor with the same shape as input.
+
+    Examples:
+        .. code-block:: python
+            :name: RReLU-example
+
+            import paddle
+            import numpy as np
```
Comment (Ligoml) [inline python/paddle/nn/layer/activation.py]: 代码示例中不需要import numpy，可以使用paddle.to_tensor来生成输入数据

File: python/paddle/nn/functional/activation.py (hunk 9 above)
Comment (thunder95) [reply python/paddle/nn/functional/activation.py]: 尝试着修改了一下，但不知道怎么预览，如果可以就先自己检查了。

File: python/paddle/nn/layer/activation.py (hunk 10 above)
Comment (thunder95) [reply python/paddle/nn/layer/activation.py]: 已修改

File: python/paddle/nn/functional/activation.py (hunk 9 above)
Comment (Ligoml) [reply python/paddle/nn/functional/activation.py]: 中英文预览都在docs的pr里面，可以参考[文档](https://github.com/PaddlePaddle/docs/wiki/%5BBeta%5D%E9%A3%9E%E6%A1%A8%E6%96%87%E6%A1%A3%E9%A2%84%E8%A7%88%E5%B7%A5%E5%85%B7)

Comment (Ligoml) [reply python/paddle/nn/functional/activation.py]: 英文的公式还是稍微有点显示错误

File: python/paddle/nn/layer/activation.py [hunk 11]
```diff
@@ -436,6 +436,72 @@ def extra_repr(self):
             name_str)
 
 
+class RReLU(Layer):
+    """
+    rrelu activation.
+
+    `Empirical Evaluation of Rectified Activations in Convolutional Network`: https://arxiv.org/abs/1505.00853
```
Comment (Ligoml) [inline python/paddle/nn/layer/activation.py]: - 把论文的超链接加在论文标题上
- 描述里可以多加一些介绍性内容，对用户友好一些
- 英文公式显示方面还是有些问题

**Comment by zhwesky2010** [comment]
> > 1.`fix_seed`这个参数应该不需要，`seed` 也不需要，目前的2.0后的写法是直接从generator中获取seed，不需要单独去给OP设置一个seed。
> > ```
> > auto gen_cuda = ctx.GetGenerator();
> > uint64_t seed = seed_offset.first;
> > uint64_t offset = seed_offset.second;
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 2.GPU kernel 参考下 https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/exponential_op.cu#L29-L31 中的简洁写法：
> > ```
> > using MT = typename kps::details::MPTypeTrait<T>::Type;
> > funcs::uniform_distribution<MT> dist;
> > funcs::uniform_real_transform<MT> trans(min, max);
> > funcs::distribution_and_transform<T>(dev_ctx, out, dist, trans);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 调用https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/funcs/distribution_helper.h 已经封装好的函数即可。不要用thrust库，这个性能不太好。可以参考下https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/phi/kernels/gpu/poisson_kernel.cu 比较新的写法
> 
> @zhouwei25 谢谢老师指导，已完成修改意见，麻烦看看这样改是否可以。

这样实现可以的，直接调funcs里的随机数公共组件就可以了

File: python/paddle/nn/layer/activation.py (hunk 11 above)
Comment (thunder95) [reply python/paddle/nn/layer/activation.py]: @Ligoml 已修改这三点， 但是预览文档页面只找到该API的中文文档，没有这个api的英文文档， 是遗漏什么操作或是需要等？

**Comment by zhwesky2010** [comment]
LGTM

File: python/paddle/nn/layer/activation.py (hunk 11 above)
Comment (Ligoml) [reply python/paddle/nn/layer/activation.py]: 需要paddle的PR-CI-Build跑完才会有英文文档预览~

File: python/paddle/nn/layer/activation.py [hunk 12]
```diff
@@ -436,6 +436,93 @@ def extra_repr(self):
...
+        RReLU(x)=
+            \left\{
+                \begin{array}{rcl}
+                    x, & & if \ x >= 0 \\
+                    (lower + upper) * 0.5 * x, & & otherwise \\
+                \end{array}
+            \right.
+
+    where :math:`x` is the input tensor,
+    :math:`lower` and :math:`upper` are the bounds of uniform distribution.
```
Comment (Ligoml) [inline python/paddle/nn/layer/activation.py:475]: image

比之前好多了，但还是有些不太对

Comment (Ligoml) [reply python/paddle/nn/layer/activation.py:475]: 中文是正常显示的，等下我重新构建一个预览看看

Comment (thunder95) [reply python/paddle/nn/layer/activation.py:475]: 公式是从代码里直接拷贝到中文文档的，二者是一样的，不清楚为什么英文的还是有问题 @Ligoml

Comment (TCChenlong) [reply python/paddle/nn/layer/activation.py:475]: 可以试试在文档最前方加一个 “r”

**Review by Ligoml — APPROVED** [review]
LGTM for docs

**Comment by jeff41404** [comment]
the design of `paddle.nn.functional.rrelu` should also be add in [RFC](https://github.com/PaddlePaddle/community/pull/71)

**Comment by thunder95** [comment]
> the design of `paddle.nn.functional.rrelu` should also be add in [RFC](https://github.com/PaddlePaddle/community/pull/71)

@jeff41404 done.  pr link: https://github.com/PaddlePaddle/community/pull/137

**Review by jeff41404 — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py [hunk 13]
```diff
@@ -0,0 +1,326 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (chenwhql) [inline python/paddle/fluid/tests/unittests/test_rrelu_op.py:1]: 2018->2022，后面可以再追加个PR更新下

**Review by chenwhql — APPROVED** [review]

File: python/paddle/fluid/tests/unittests/test_rrelu_op.py (hunk 13 above)
Comment (thunder95) [reply python/paddle/fluid/tests/unittests/test_rrelu_op.py:1]: 好的
